File: ios/EggyRobotClient/Sources/RobotCommandEncoder.swift
import Foundation

enum RobotCommand: Equatable {
    case velocity(x: Double, y: Double, z: Double)
    case reset
    case goal(x: Double, y: Double, yaw: Double = 0)
    case autoExplore(enabled: Bool)
    case setScene(String)
    case saveMap(name: String)
    case loadMap(id: String)
    case deleteMap(id: String)
    case startMapping
    case stopMapping
    case saveNavigationMap(name: String)
    case startSimpleNav
    case stopSimpleNav
    case simpleGoal(x: Double, y: Double, yaw: Double = 0)
    case cancelSimpleGoal
    case startNavigation(mapPath: String? = nil)
    case stopNavigation
    case cancelGoal
    case ping(client: String = "ios")
}

struct RobotCommandEncoder {
    func encode(_ command: RobotCommand) -> [String: Any] {
        switch command {
        case .velocity(let x, let y, let z):
            return ["type": "cmd_vel", "linear_x": x, "linear_y": y, "angular_z": z]
        case .reset:
            return ["type": "reset"]
        case .goal(let x, let y, let yaw):
            return ["type": "goal", "frame_id": "map", "x": x, "y": y, "yaw": yaw]
        case .autoExplore(let enabled):
            return ["type": "auto_explore", "enabled": enabled]
        case .setScene(let id):
            return ["type": "set_scene", "scene": id]
        case .saveMap(let name):
            return ["type": "save_map", "name": name]
        case .loadMap(let id):
            return ["type": "load_map", "id": id]
        case .deleteMap(let id):
            return ["type": "delete_map", "id": id]
        case .setMode(let mode):
            return ["type": "set_mode", "mode": mode]
        case .startMapping:
            return ["type": "start_mapping"]
        case .stopMapping:
            return ["type": "stop_mapping"]
        case .saveNavigationMap(let name):
            return ["type": "save_navigation_map", "name": name]
        case .startSimpleNav:
            return ["type": "start_simple_nav"]
        case .stopSimpleNav:
            return ["type": "stop_simple_nav"]
        case .simpleGoal(let x, let y, let yaw):
            return ["type": "simple_goal", "frame_id": "map", "x": x, "y": y, "yaw": yaw]
        case .cancelSimpleGoal:
            return ["type": "cancel_simple_goal"]
        case .startNavigation(let mapPath):
            var payload: [String: Any] = ["type": "start_navigation"]
            if let mapPath, !mapPath.isEmpty { payload["map_path"] = mapPath }
            return payload
        case .stopNavigation:
            return ["type": "stop_navigation"]
        case .cancelGoal:
            return ["type": "cancel_goal"]
        case .ping(let client):
            return ["type": "ping", "client": client]
        }
    }
}

File: android/EggyRosCar/app/src/main/java/com/liueggy/roscarapp/data/RobotCommandEncoder.kt
package com.liueggy.roscarapp.data

sealed class RobotCommand {
    data class Velocity(val x: Double, val y: Double, val z: Double) : RobotCommand()
    data object Reset : RobotCommand()
    data class Goal(val x: Double, val y: Double, val yaw: Double = 0.0) : RobotCommand()
    data class AutoExplore(val enabled: Boolean) : RobotCommand()
    data class SetScene(val id: String) : RobotCommand()
    data class SaveMap(val name: String) : RobotCommand()
    data class LoadMap(val id: String) : RobotCommand()
    data class DeleteMap(val id: String) : RobotCommand()
    data class SetMode(val mode: String) : RobotCommand()
    data class Ping(val client: String = "android") : RobotCommand()
}

class RobotCommandEncoder {
    fun encode(command: RobotCommand): Map<String, Any> = when (command) {
        is RobotCommand.Velocity -> mapOf("type" to "cmd_vel", "linear_x" to command.x, "linear_y" to command.y, "angular_z" to command.z)
        RobotCommand.Reset -> mapOf("type" to "reset")
        is RobotCommand.Goal -> mapOf("type" to "goal", "frame_id" to "map", "x" to command.x, "y" to command.y, "yaw" to command.yaw)
        is RobotCommand.AutoExplore -> mapOf("type" to "auto_explore", "enabled" to command.enabled)
        is RobotCommand.SetScene -> mapOf("type" to "set_scene", "scene" to command.id)
        is RobotCommand.SaveMap -> mapOf("type" to "save_map", "name" to command.name)
        is RobotCommand.LoadMap -> mapOf("type" to "load_map", "id" to command.id)
        is RobotCommand.DeleteMap -> mapOf("type" to "delete_map", "id" to command.id)
        is RobotCommand.SetMode -> mapOf("type" to "set_mode", "mode" to command.mode)
        is RobotCommand.Ping -> mapOf("type" to "ping", "client" to command.client)
    }
}
